File: backend/libexecution/libjson.ml
open Core_kernel
open Lib
open Types.RuntimeT
module RT = Runtime

let fns : fn list =
  [ { prefix_names = ["JSON::read"]
    ; infix_names = []
    ; parameters = [par "json" TStr]
    ; return_type = TAny
    ; description =
        "Parses a json string and returns its value. HTTPClient functions, and our request handler, automatically parse JSON into the `body` and `jsonbody` fields, so you probably won't need this. However, if you need to consume bad JSON, you can use string functions to fix the JSON and then use this function to parse it."
    ; func =
        InProcess
          (function
          | _, [DStr json] ->
            ( try Dval.of_unknown_json_v0 (Unicode_string.to_string json)
              with _ -> DNull )
          | args ->
              fail args)
    ; preview_execution_safe = true
    ; deprecated = true }
  ; { prefix_names = ["JSON::read_v1"]
    ; infix_names = []
    ; parameters = [par "json" TStr]
    ; return_type = TAny
    ; description =
        "Parses a json string and returns its value. HTTPClient functions, and our request handler, automatically parse JSON into the `body` and `jsonbody` fields, so you probably won't need this. However, if you need to consume bad JSON, you can use string functions to fix the JSON and then use this function to parse it."
    ; func =
        InProcess
          (function
          | _, [DStr json] ->
              Dval.of_unknown_json_v1 (Unicode_string.to_string json)
          | args ->
              fail args)
    ; preview_execution_safe = true
    ; deprecated = true }
  ; { prefix_names = ["JSON::parse_v0"]
    ; infix_names = []
    ; parameters = [par "json" TStr]
    ; return_type = TAny
    ; description =
        "Parses a json string and returns its value. HTTPClient functions, and our request handler, automatically parse JSON into the `body` and `jsonbody` fields, so you probably won't need this. However, if you need to consume bad JSON, you can use string functions to fix the JSON and then use this function to parse it."
    ; func =
        InProcess
          (function
          | _, [DStr json] ->
              Dval.of_unknown_json_v1 (Unicode_string.to_string json)
          | args ->
              fail args)
    ; preview_execution_safe = true
    ; deprecated = true }
  ; { prefix_names = ["JSON::parse_v1"]
    ; infix_names = []
    ; parameters = [par "json" TStr]
    ; return_type = TResult
    ; description =
        "Parses a json string and returns its value. HTTPClient functions, and our request handler, automatically parse JSON into the `body` and `jsonbody` fields, so you probably won't need this. However, if you need to consume bad JSON, you can use string functions to fix the JSON and then use this function to parse it."
    ; func =
        InProcess
          (function
          | _, [DStr json] ->
            ( try
                let dval =
                  Dval.of_unknown_json_v1 (Unicode_string.to_string json)
                in
                DResult (ResOk dval)
              with e ->
                DResult
                  (ResError
                     (e |> Exception.exn_to_string |> Dval.dstr_of_string_exn))
            )
          | args ->
              fail args)
    ; preview_execution_safe = true
    ; deprecated = false } ]

File: fsharp-backend/src/LibExecution/StdLib/LibJson.fs
module LibExecution.StdLib.LibJson

open System.Threading.Tasks
open System.Numerics
open FSharp.Control.Tasks
open FSharpPlus

open LibExecution.RuntimeTypes
open Prelude

module Errors = LibExecution.Errors
module DvalRepr = LibExecution.DvalRepr

let fn = FQFnName.stdlibFnName

let err (str : string) = Value(Dval.errStr str)

let incorrectArgs = LibExecution.Errors.incorrectArgs

let varErr = TVariable "err"
let varA = TVariable "a"
let varB = TVariable "b"

let fns : List<BuiltInFn> =
  [ { name = fn "JSON" "read" 0
      parameters = [ Param.make "json" TStr "" ]
      returnType = varA
      description =
        "Parses a json string and returns its value. HTTPClient functions, and our request handler, automatically parse JSON into the `body` and `jsonbody` fields, so you probably won't need this. However, if you need to consume bad JSON, you can use string functions to fix the JSON and then use this function to parse it."
      fn =
        (function
        | _, [ DStr json ] ->
            (try
              json |> DvalRepr.ofUnknownJsonV0 |> Value
             with _ -> Value DNull)
        | _ -> incorrectArgs ())
      sqlSpec = NotYetImplementedTODO
      previewable = Pure
      deprecated = ReplacedBy(fn "JSON" "read" 1) }
    { name = fn "JSON" "read" 1
      parameters = [ Param.make "json" TStr "" ]
      returnType = varA
      description =
        "Parses a json string and returns its value. HTTPClient functions, and our request handler, automatically parse JSON into the `body` and `jsonbody` fields, so you probably won't need this. However, if you need to consume bad JSON, you can use string functions to fix the JSON and then use this function to parse it."
      fn =
        (function
        | _, [ DStr json ] -> json |> DvalRepr.ofUnknownJsonV1 |> Value
        | _ -> incorrectArgs ())
      sqlSpec = NotYetImplementedTODO
      previewable = Pure
      deprecated = ReplacedBy(fn "JSON" "parse" 0) }
    { name = fn "JSON" "parse" 0
      parameters = [ Param.make "json" TStr "" ]
      returnType = varA
      description =
        "Parses a json string and returns its value. HTTPClient functions, and our request handler, automatically parse JSON into the `body` and `jsonbody` fields, so you probably won't need this. However, if you need to consume bad JSON, you can use string functions to fix the JSON and then use this function to parse it."
      fn =
        (function
        | _, [ DStr json ] ->
            try
              json |> DvalRepr.ofUnknownJsonV1 |> Value
            with e -> Value(DError(SourceNone, e.Message))
        | _ -> incorrectArgs ())
      sqlSpec = NotYetImplementedTODO
      previewable = Pure
      deprecated = ReplacedBy(fn "JSON" "parse" 1) }
    { name = fn "JSON" "parse" 1
      parameters = [ Param.make "json" TStr "" ]
      returnType = TResult(varA, varErr)
      description =
        "Parses a json string and returns its value. HTTPClient functions, and our request handler, automatically parse JSON into the `body` and `jsonbody` fields, so you probably won't need this. However, if you need to consume bad JSON, you can use string functions to fix the JSON and then use this function to parse it."
      fn =
        (function
        | _, [ DStr json ] ->
            (try
              let dval = json |> DvalRepr.ofUnknownJsonV1
              Value(DResult(Ok dval))
             with Failure (e) -> e.ToString() |> DStr |> Error |> DResult |> Value)
        | _ -> incorrectArgs ())
      sqlSpec = NotYetImplementedTODO
      previewable = Pure
      deprecated = NotDeprecated } ]
